File: 프로그래머스/3/181186. 아방가르드 타일링/아방가르드 타일링.cpp
// https://ampersandor.tistory.com/21
#include <string>
#include <vector>

using namespace std;

const int MOD = 1000000007;

int solution(int n) {
    vector<long long> dp(100001);
    dp[0] = 0;
    dp[1] = 1;
    dp[2] = 3;
    dp[3] = 10;
    dp[4] = 23;
    dp[5] = 62;
    dp[6] = 170;
    
    if (n <= 6) return dp[n];
    
    for (int i = 7; i <= n; i++) {
        dp[i] = (dp[i-1] + 2 * dp[i-2] + 6 * dp[i-3] + dp[i-4] - dp[i-6]) % MOD;
        if (dp[i] < 0) {
            dp[i] += MOD;
        }
    }
    
    return (int)dp[n];
}

File: 프로그래머스/3/181186. 아방가르드 타일링/아방가르드 타일링.java
class Solution {
    static final long MOD = 1_000_000_007L;

    public int solution(int n) {
        long[] dp = new long[n + 1];
        
        dp[0] = 1;
        if (n >= 1) dp[1] = 1;
        if (n >= 2) dp[2] = 3;
        if (n >= 3) dp[3] = 10;
        
        long[] sum = new long[3];
        
        for (int i = 4; i <= n ; i++) {
            sum[(i - 4) % 3] = (sum[(i - 4) % 3] + dp[i - 4]) % MOD;
            
            dp[i] = (
                    dp[i - 1]
                    + 2 * dp[i - 2]
                    + 5 * dp[i - 3]
                    + 2 * (sum[0] + sum[1] + sum[2])
                    + 2 * sum[i % 3]
            ) % MOD;
        }
        
        return (int) dp[n];
    }
}